app/models/payment_intent.rb

```diff
@@ -0,0 +1,153 @@
+  }
+
+  def started?
+    self.wca_status != 'created'
+  end
+
+  def retrieve_intent
+    if payment_record_type == 'StripeRecord'
+      Stripe::PaymentIntent.retrieve(
+        self.stripe_id,
+        client_secret: client_secret,
+        stripe_account: self.find_account_id,
+      )
+    else
+      raise "Trying to call retrieve_intent for a PaymentIntent with unmatched payment_record_type of: #{payment_record_type}"
+    end
+  end
+
+  def update_status_and_charges(api_intent, action_source, source_datetime = DateTime.current, &block)
+    if payment_record_type == 'StripeRecord'
+      update_stripe_status_and_charges(api_intent, action_source, source_datetime, &block)
+    else
+      raise "Trying to update status and charges for a PaymentIntent with unmatched payment_record_type of: #{payment_record_type}"
+    end
+  end
+
+  def update_stripe_status_and_charges(api_intent, action_source, source_datetime)
+    ActiveRecord::Base.transaction do
+      self.update!(error_details: api_intent.last_payment_error)
+      self.payment_record.assign_status(api_intent) # Will also update the wca_status
+
+      # Payment Intent lifecycle as per https://stripe.com/docs/payments/intents#intent-statuses
+      case api_intent.status
+      when 'succeeded'
+        # The payment didn't need any additional actions and is completed!
+
+        # Record the success timestamp if not already done
+        unless self.succeeded?
+          self.assign_attributes(
+            confirmed_at: source_datetime,
+            confirmation_source: action_source,
+            wca_status: payment_record.determine_wca_status,
+          )
+          save
+          payment_record.save
+        end
+
+        intent_charges = Stripe::Charge.list(
+          { payment_intent: self.stripe_id },
+          stripe_account: self.find_account_id,
+        )
+
+        intent_charges.data.each do |charge|
+          recorded_transaction = StripeRecord.find_by(stripe_id: charge.id)
+
+          if recorded_transaction.present?
+            recorded_transaction.update_status(charge)
+          else
+            fresh_transaction = StripeRecord.create_from_api(charge, {})
+            fresh_transaction.update!(parent_transaction: self.payment_record)
+
+            # Only trigger outer update blocks for charges that are actually successful. This is reasonable
+            # because we only ever trigger this block for PIs that are marked "successful" in the first place
+            charge_successful = fresh_transaction.stripe_status == "succeeded"
+
+            yield fresh_transaction if block_given? && charge_successful
+          end
+        end
+      when 'canceled'
+        # Canceled by Stripe
+        self.update!(
+          canceled_at: source_datetime,
+          cancellation_source: action_source,
+          wca_status: payment_record.determine_wca_status,
+        )
+        save
+        payment_record.save
```

Comment:
(Very probably) redundant
`update!` already fires the SQL call, and saving the associated `payment_record` seems unnecessary.

Reply:
See above